src/java.base/windows/native/libjava/WinNTFileSystem_md.c

```diff
@@ -351,6 +331,25 @@ Java_java_io_WinNTFileSystem_canonicalizeWithPrefix0(JNIEnv *env, jobject this,
     return rv;
 }
 
+JNIEXPORT jstring JNICALL
+Java_java_io_WinNTFileSystem_getFinalPath0(JNIEnv* env, jobject this, jstring pathname) {
+    jstring rv = NULL;
+
+    WITH_UNICODE_STRING(env, pathname, path) {
+        WCHAR* finalPath = getFinalPath(env, path);
+        if (finalPath != NULL) {
+            rv = (*env)->NewString(env, finalPath, (jsize)wcslen(finalPath));
+            free(finalPath);
+        }
+    } END_UNICODE_STRING(env, path);
+
+    if (rv == NULL && !(*env)->ExceptionCheck(env)) {
+        JNU_ThrowIOExceptionWithLastError(env, "Bad pathname"); // XXX message?
```

Comment:
That "not useful" exception message is probably okay because this is what canonicalize has always used.

Reply:
Removed "XXX" note in 796a634.